crates/ty_python_semantic/src/types.rs

```diff
@@ -6722,18 +6722,14 @@ impl<'db> Type<'db> {
                     invalid_expressions: smallvec::smallvec_inline![InvalidTypeExpression::Generic],
                     fallback_type: Type::unknown(),
                 }),
-                KnownInstanceType::UnionType(list) => {
-                    let mut builder = UnionBuilder::new(db);
-                    let inferred_as = list.inferred_as(db);
-                    for element in list.elements(db) {
-                        builder = builder.add(if inferred_as.type_expression() {
-                            *element
-                        } else {
-                            element.in_type_expression(db, scope_id, typevar_binding_context)?
-                        });
+                KnownInstanceType::UnionType(list) => match list {
+                    UnionTypeInstance::PEP604(instance) => {
+                        // Cloning here is cheap if the result is a `Type` (which is `Copy`). It's more
+                        // expensive if there are errors.
+                        instance.union_type(db).clone()
```

Comment:
@carljm In the end, this didn't require a new salsa query. Instead, we're now eagerly building that `Type::Union` in case there were no errors, and interning the result. And then we only need to copy the result here when we see a use of this implicit type alias in a type expression.

This means that we do some extra work if someone defines `IntOrStr = int | str` and then never uses it as a type. But that feels okay to me?